src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,348 @@
+            {
+                return compilationUnit.Members;
+            }
+
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        /// <summary>
+        /// Try to get a new node to replace given node, which is a reference to a top-level type declared inside the namespce to be changed.
+        /// If this reference is the right side of a qualified name, the new node returned would be the entire qualified name. Depends on 
+        /// whether <paramref name="newNamespaceParts"/> is provided, the name in the new node might be qualified with this new namespace instead.
+        /// </summary>
+        /// <param name="reference">A reference to a type declared inside the namespce to be changed, which is calculated based on results from 
+        /// `SymbolFinder.FindReferencesAsync`.</param>
+        /// <param name="newNamespaceParts">If specified, and the reference is qualified with namespace, the namespace part of original reference 
+        /// will be replaced with given namespace in the new node.</param>
+        /// <param name="old">The node to be replaced. This might be an ancestor of original reference.</param>
+        /// <param name="new">The replacement node.</param>
+        public override bool TryGetReplacementReferenceSyntax(
+            SyntaxNode reference,
+            ImmutableArray<string> newNamespaceParts, 
+            ISyntaxFactsService syntaxFacts, 
+            out SyntaxNode old, 
+            out SyntaxNode @new)
+        {
+            if (!(reference is SimpleNameSyntax nameRef))
+            {
+                old = @new = null;
+                return false;
+            }
+
+            // A few different cases are handled here:
+            //
+            // 1. When the reference is not qualified (i.e. just a simple name), then there's nothing need to be done.
+            //    And both old and new will point to the original reference.
+            //
+            // 2. When the new namespace is not specified, we don't need to change the qualified part of reference.
+            //    Both old and new will point to the qualified reference.
+            //
+            // 3. When the new namespace is "", i.e. we are moving type referenced by name here to global namespace.
+            //    As a result, we need replace qualified reference with the simple name.
+            //
+            // 4. When the namespace is specified and not "", i.e. we are moving referenced type to a different non-global 
+            //    namespace. We need to replace the qualified reference with a new qualified reference (which is qualified 
+            //    with new namespace.)
+                        
+            if (syntaxFacts.IsRightSideOfQualifiedName(nameRef))
+            {
+                old = nameRef.Parent;
+                var aliasQualifier = GetAliasQualifierOpt(old);
+
+                if (IsGlobalNamespace(newNamespaceParts))
```

Comment:
is there any reason this doesn't just use expander ? 
(http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Simplification/Simplifier.cs,71)

I saw the test that things just got simplified at the end, so, rather than having custom logic to have fully qualified node in some cases and others, just expanding the node at the reference context and let simplifier at the end take cares of it might be simpler for you?